packages/algob/test/internal/deployer.ts

```diff
@@ -676,4 +677,74 @@ describe("DeployerDeployMode", () => {
 		assert.isDefined(res);
 		assert.deepEqual(res.assetDef, asaDef);
 	});
+
+	describe("addCheckpointApp and addCheckpointASA", function () {
+		let deployer: DeployerDeployMode;
+
+		this.beforeEach(() => {
+			deployer = new DeployerDeployMode(deployerCfg);
+		});
+
+		it("addCheckpointApp", () => {
+			const appInfo = {
+				creator: algosdk.encodeAddress(mockConfirmedTx.txn.txn.snd),
+				applicationAccount: algosdk.getApplicationAddress(mockConfirmedTx["application-index"]),
+				txID: algosdk.Transaction.from_obj_for_encoding(mockConfirmedTx.txn.txn).txID(),
+				confirmedRound: 1,
+				appID: mockConfirmedTx["application-index"],
+				timestamp: 1,
+				deleted: false,
+				approvalFile: "YXBwcm92YWw=",
+				clearFile: "Y2xlYXI=",
+			};
+
+			deployer.addCheckpointApp("app", mockConfirmedTx);
+
+			// get checkpoint data
+			const checkpointData = deployerCfg.cpData.precedingCP["network 123"].app
+				.get("app")
+				?.values()
+				.next().value;
+
+			checkpointData.timestamp = 1; // sync timestamp
+
+			assert.deepEqual(checkpointData, appInfo);
+
+			assert.deepEqual(appInfo, deployer.getApp("app"));
+		});
+
+		it("addCheckpointASA", () => {
+			const asaInfo: rtypes.ASAInfo = {
+				assetDef: { decimals: 0, total: 10 },
+				assetIndex: 1,
+				creator: algosdk.encodeAddress(mockConfirmedTx.txn.txn.snd),
+				deleted: false,
+				confirmedRound: 1,
+				txID: algosdk.Transaction.from_obj_for_encoding(mockConfirmedTx.txn.txn).txID(),
+			};
+
+			deployer.addCheckpointASA("asaName", mockConfirmedTx);
+
+			// get checkpoint data
+			const checkpointData = deployerCfg.cpData.precedingCP["network 123"].asa.get("asaName");
+
+			assert.deepEqual(checkpointData, asaInfo);
+
+			assert.deepEqual(asaInfo, deployer.getASAInfo("asaName"));
+		});
```

Comment:
Can we also add tests for deploying multiple apps/asa's ?

Reply:
Yeah, sure! I am missing the test for dup checkpoint. 😅